helix-core/src/test/java/org/apache/helix/integration/task/TestForceDeleteWorkflow.java

```diff
@@ -0,0 +1,433 @@
+  protected final String CLUSTER_NAME = CLUSTER_PREFIX + "_" + getShortClassName();
+  protected MockParticipantManager[] _participants;
+
+  protected ZkHelixClusterVerifier _clusterVerifier;
+  private HelixAdmin _admin;
+
+  @BeforeClass
+  public void beforeClass() throws Exception {
+    super.beforeClass();
+    _participants = new MockParticipantManager[_numNodes];
+    _gSetupTool.addCluster(CLUSTER_NAME, true);
+
+    setupParticipants(_gSetupTool);
+    setupDBs();
+
+    for (int i = 0; i < _numNodes; i++) {
+      startParticipant(ZK_ADDR, i);
+    }
+
+    _manager = HelixManagerFactory.getZKHelixManager(CLUSTER_NAME, "Admin",
+        InstanceType.ADMINISTRATOR, ZK_ADDR);
+    _manager.connect();
+    _driver = new TaskDriver(_manager);
+
+    _clusterVerifier =
+        new BestPossibleExternalViewVerifier.Builder(CLUSTER_NAME).setZkClient(_gZkClient).build();
+
+    // start controller
+    String controllerName = CONTROLLER_PREFIX + "_0";
+    _controller = new ClusterControllerManager(ZK_ADDR, CLUSTER_NAME, controllerName);
+    _controller.syncStart();
+
+    Assert.assertTrue(_clusterVerifier.verifyByPolling());
+
+    _admin = _gSetupTool.getClusterManagementTool();
+  }
+
+  @AfterClass
+  public void afterClass() throws Exception {
+    if (_controller != null && _controller.isConnected()) {
+      _controller.syncStop();
+    }
+    if (_manager != null && _manager.isConnected()) {
+      _manager.disconnect();
+    }
+    for (int i = 0; i < _numNodes; i++) {
+      stopParticipant(i);
+    }
+    deleteCluster(CLUSTER_NAME);
+  }
+
+  protected void setupDBs() {
+    setupDBs(_gSetupTool);
+  }
+
+  protected void setupDBs(ClusterSetup clusterSetup) {
+    // Set up target db
+    boolean partitionVary = true;
+    if (_numDbs > 1) {
+      for (int i = 0; i < _numDbs; i++) {
+        int varyNum = partitionVary ? 10 * i : 0;
+        String db = WorkflowGenerator.DEFAULT_TGT_DB + i;
+        clusterSetup.addResourceToCluster(CLUSTER_NAME, db, _numPartitions + varyNum,
+            MASTER_SLAVE_STATE_MODEL, IdealState.RebalanceMode.FULL_AUTO.toString());
+        clusterSetup.rebalanceStorageCluster(CLUSTER_NAME, db, _numReplicas);
+        _testDbs.add(db);
+      }
+    } else {
+      clusterSetup.addResourceToCluster(CLUSTER_NAME, WorkflowGenerator.DEFAULT_TGT_DB,
+          _numPartitions, MASTER_SLAVE_STATE_MODEL, IdealState.RebalanceMode.FULL_AUTO.name());
+      clusterSetup.rebalanceStorageCluster(CLUSTER_NAME, WorkflowGenerator.DEFAULT_TGT_DB,
+          _numReplicas);
+    }
+  }
+
+  protected void setupParticipants(ClusterSetup setupTool) {
+    _participants = new MockParticipantManager[_numNodes];
+    for (int i = 0; i < _numNodes; i++) {
+      String storageNodeName = PARTICIPANT_PREFIX + "_" + (_startPort + i);
+      setupTool.addInstanceToCluster(CLUSTER_NAME, storageNodeName);
+    }
+  }
+
+  protected void startParticipant(String zkAddr, int i) {
```

Comment:
Is this a copy-paste? Please avoid having duplicate code. https://dzone.com/articles/software-design-principles-dry-and-kiss "DRY principle".